### ISSUE TITLE
Reset in integration testing should clear styles

### ISSUE THREAD

**Issue by nathanmarks** [cssinjs/jss#293]
Right now, tests end up polluting the document head with style tags and they are not cleaned up between tests.

I added code to `reset()` to remove the `jss` stylesheets, however, this ends up breaking a bunch of tests because they are expecting the DOM elements to exist despite the fact that `reset()` is being called `afterEach()` test.

Performing `reset()` `afterEach()` test is not ideal because it prevents using several `it()` blocks between resets. The way `reset()` being used should be changed as it would also be ideal to clear the `style` tags. Without that it's hard to do any integration testing involving assertions on the DOM output. It seems odd that making `reset()` clear style tags ends up breaking tests as it clears the sheet and plugin registries for the JSS instance.

Related note: I noticed that `afterEach()` is called at the top of each test file. This will result in the `afterEach()` calls being stacked by mocha as they are not nested inside the `describe()` blocks for each test file.

**Comment by kof** [comment]
I thought all tests which call .attach() also call .detach() ... need to go through them and find the missing once.

**Comment by kof** [comment]
> Related note: I noticed that afterEach() is called at the top of each test file. This will result in the afterEach() calls being stacked by mocha as they are not nested inside the describe() blocks for each test file.

So we need to put afterEach into corresponding describe scope ...

**Comment by kof** [comment]
> Performing reset() afterEach() test is not ideal because it prevents using several it() blocks between resets. The way reset() being used should be changed as it would also be ideal to clear the style tags. Without that it's hard to do any integration testing involving assertions on the DOM output. It seems odd that making reset() clear style tags ends up breaking tests as it clears the sheet and plugin registries for the JSS instance.

Can you provide a specific test case to make sure we are on the same page?

**Comment by kof** [comment]
Found only one place where a sheet is not detached after being attached, @nathanmarks do you see more?

**Comment by nathanmarks** [comment]
@kof Nope, I only saw 1 extra in the DOM at the start of the tests I was writing.

When you `reset()`, wouldn't it also make sense to detach stylesheets since you're clearing the registry? Either way, however it gets implemented -- it would be nice to have a cleanup routine at the end of each test (or start of each) module so that other, independent tests don't have a chance of getting screwed by a test not doing proper cleanup. It can be frustrating to debug in a big test suite depending on the issue.

**Comment by kof** [comment]
We can add  something like that, just to be sure: `jss.sheets.registry.forEach(jss.removeStyleSheet, jss)`

**Comment by kof** [comment]
I think what you are referring to is the use of before() and after() ... where the it() calls in between rely on them. There are situations where you want to have a number of it() calls without to clean up in between, because you need multiple steps to fully describe the scenario.

Maybe I just misunderstand what you mean ....

**Comment by kof** [comment]
I think I found a way, moving towards to having fully isolated it() calls everywhere, will be done in a sec.

**Comment by kof** [comment]
@nathanmarks please check if I covered everything ... if not please reopen.
